### ISSUE TITLE
Create file reference functionality such as holding reference iRODS files

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#2343]
To serve two HydroShare-based use cases at RENCI, after some discussions with @mjstealey and @cbcunc, I decided it is better to just create a new iRODS reference resource type in HydroShare code base as the basis which can be directly leveraged by those two use cases at RENCI as well as by HydroShare community. In both use cases at RENCI, their data are already in their own iRODS stores, and they want to use HydroShare to reference those data as the front end user interface without moving data into HydroShare iRODS store/zone. This issue will involve creating a new generic iRODS reference resource type that references data stored in other iRODS servers which HydroShare  iRODS proxy user does not have direct access to. All AVU metadata stored on IRODS servers associated with the referenced data will populate as extended key-value metadata on Django side for the new iRODS reference resource type, and users can add and edit metadata as needed, but users cannot download the referenced files. The referenced file entries cannot be directly edited to protect the reference path integrity, but the file entries can be deleted or new file entries can be added via iRODS file browser. The downloaded bag will include metadata and resourcemap xml files and a referenced_files.txt file under data/contents that includes paths for all referenced files in this iRODS reference resource. 

I also discussed specifics with @mjstealey on leveraging this new iRODS reference resource type for nwm data. We both think a NWM reference resource type could be created as a sub-class to the generic iRODS reference resource type, and can handle updating referenced nwm data path in the resource based on the sliding data refresh window. 

Since one use case at RENCI that will leverage this new reference resource type is for SC17 demo which will take place on Nov 12, I will have to work on this as a priority item to make the time line. I may not get the full feature set done for this new resource type by Nov. 3 when we plan to do integration testing for SC17 demo,  but can work on this further after SC17 demo to get it ready for merging into HydroShare. We think it is a useful feature to get into HydroShare, but @dtarb let us know your thought on this.

**Comment by jonc1438** [comment]
This is a little self-serving but sound great to me since we are one of those use cases!

One question/issue since the metadata in AVUs will sync with Django metadata; 
If metadata is edited on the Django side will that get pushed back to iRODS AVUs to keep both in sync? 

Another edge case would be the reverse where metadata AVU's in iRODS altered after reference ingest requiring re-ingest or refresh of Django metadata.

Ignore this newbie's question if it does not make sense.

**Comment by hyi** [comment]
@jonc1438 Your questions make perfect sense and the answers to both questions are no since it is a reference resource that reflects the snapshot/status of the data at the creation time, and after the resource is created, HydroShare will not have the iRODS user credential to access the referenced data to do any update. However, the user can delete the old file entries, and add the new file entries via iRODS file browser from resource landing page, and this process will bring the referenced data in sync with iRODS back end again. In summary, HydroShare cannot keep things sync'ed up automatically, but provide means for users to do the sync-up as needed. This is how a generic iRODS reference resource would behave. You can definitely create a subclass to this generic iRODS reference resource to add this automatic sync-up for your specific workflow as needed. @mjstealey please add anything I have missed to clarify things as needed.

**Comment by horsburgh** [comment]
@hyi - would the content have to be stored in iRODS? For example - would this new referenced resource type be usable for referencing a file that is available using an HTTP URL for instance?

Or - are we talking about something completely different here?

**Comment by mjstealey** [comment]
@jonc1438 - What @hyi is scoping is the basis for an **iRODS reference resource** which can be extended to suit different use cases as needed by other developers. The initial pass will serve as a means of registering an iRODS file or collection, but may not check all the boxes that all users want.

Hong will however have encoded a template for others to build upon.

### Basis
- **iRODS reference resource**
	- iRODS collection/file information
		- Size, zone location, etc.
		- AVUs
	- Additional user added metadata
		- Contact information
		- iRODS credential info.
		- information about the resource not encapsulated in the reference files themselves
		- etc.
	- Assumptions
		- actual files will not reside in HydroShare
		- user is responsible for managing state of reference if underlying iRODS files change (DOI implications here)
		- user may not add additional static files to the resource, it is simply a reference to things in iRODS in this basic form

Other work can then be extended from this, for example:

### Extends
- **NWM iRODS reference resource**
	- All of **iRODS reference resource** +
	- accomodate a sliding window of time that the resource exists in iRODS
		- `analysis_assim` - 720 day window
		- `short_range` - 40 day window
		- etc.

- **Brain-I iRODS reference resource**
	- All of **iRODS reference resource** +
		- AVU validation / re-ingest services
			- check AVU state based on time modified for Django vs iRODS
		- workflow and automation information
		- specific apps that operate on this resource type
		- etc.

**Comment by dtarb** [comment]
I have the following comments and concerns
1.	I am concerned about prioritizing this for use cases that have not been described.  Are there written descriptions for the use cases that we can discuss to understand the best way to solve them.
2.	There are AVU details in the first paragraph that are stated without context.  We need the bigger picture to get a sense of why AVU’s are involved.  I would prefer that AVU’s not be involved as they seem limiting.
3.	We should do our best to keep this simple.  If content is to be held outside of HydroShare just use files in a composite resource to describe how to access those files.  That is what a reference is – a description of how to access something outside.  
4.	If there is to be file specific metadata it should be associated with a “file type” not a resource type.  We should be using this model for all new “types” in HydroShare so as to be able to take advantage of the flexibility due to decoupling “resource” from “type” that composite resources give us.
5.	I suggest that for a first implementation (and SC17 seems like a first implementation) just work with files in a composite resource without trying to rush an additional type into the code.  Use this working with files to gain understanding of use patterns and the requirements for type specific metadata to be added once requirements are understood.
6.	There are many examples of things that need to be referenced from HydroShare.  A model we develop for reference resources should recognize this and to the extent possible use the same approach for all.  We already have reference HIS time series types in composite resources.  We could go to referencing external content in an iRODS server somewhere, referencing external content on a website somewhere (http server), referencing external content behind various types of API’s such as OGC web services, and I am sure others.  I think we should look for a general way for all of these could be accommodated by a file that provides information on how to access the external resource.

**Comment by horsburgh** [comment]
@dtarb - one of the purposes behind the way we encoded the science metadata and resource map for a resource is that each "file" that comprises a resource is referenced using a URI. So, whether the file exists in iRODS or not, that URI should point at the file. Technically, the system should be able to read the resource map (and/or its Django reflection) and then display a list of all of the files in the resource regardless of where they are physically located. Where I am going with this, is that I don't think we need "a file that provides information on how to access the external resource" - we already have a way to do this in our science metadata and resource map.  What we need are ways to add referenced content to a resource without uploading a file (e.g., specifying a link instead).

Apologies if I'm hijacking this issue and conflating multiple things.

**Comment by dtarb** [comment]
@horsburgh I was following reference time series file type as the example.  If there is a better way to do it I am happy for you and others on the data model team to work it out.  My suggestion is that we still need to articulate the design and purpose for what is being done here (unless this is in documents I have not seen) and look for a simple solution that is general for quite a number of kinds of content we may want to reference.  I am worried there is too much iRODS specific AVU complexity in what seems to be proposed (but exactly what is being proposed is unclear to me).  We need better context and discussion to sort this out.

**Comment by pkdash** [comment]
Someone should start writing a specification for a generic reference type (file/resource) in HydroShare which should help if needed writing more specific reference type like iRODS. Only then think about implementation.

**Comment by mjstealey** [comment]
@all -this  topic is an agenda item for team call on 10/11/2017

**Comment by hyi** [comment]
@dtarb @horsburgh Just got a chance to read through details of your comments on this issue carefully and give this some more thought. I think I agree every comment both of you made. Let me put another option on the table after giving this some more thought (sorry this issue caught you off-guard, but this idea just came about yesterday afternoon triggered by the pressing need for SC17 demo). Instead of creating a new reference resource type, to serve the two specific use cases (we can discuss more on the use cases in our team call today), we just need our ResourceFile object in Django to support a referenced file by an iRODS logical path or a URL (as brought up by @horsburgh), then on iRODS file browser in HydroShare, I can add two check boxes to allow users to check to indicate whether they want the iRODS files as reference in HydroShare rather than physically copied over to HydroShare, and whether they want AVUs to be copied over as key-value pair metadata in HydroShare. I will start out to only support the first checkbox that allow users to check to indicate they want iRODS files selected to be reference rather than the default copy-over behavior, and leave AVU checkbox later or just defer to @jonc1438's team to implement for their own customized HydroShare instance to fit their use case which can be feed back into HydroShare or xDCIShare, whichever is appropriate. The high-level list of items to be done for this option:
- modifying underlying ResourceFile class in Django to support referenced file field model. 
- add checkbox in iRODS file browser to allow users to check to indicate they want selected iRODS files to be referenced rather than copied over.
- update file browser on resource landing page to support display and deletion of the referenced files.

I can take care of the first two items and may need help for the last item from @Maurier to get it polished and work well. For the SC17 demo, I can make it work as simple as possible just for demo purpose due to pressing deadline. I think this option would be sufficient to address the use cases, without the complexity a generic reference resource type would incur from HydroShare design perspective. We can discuss further in our call today.

**Comment by hyi** [comment]
@horsburgh @dtarb I am working on this to create a reference file type for SC17 demo, and have question on how we should encode the referenced file URI in the resource bag. In this SC17 demo case, the reference file path is the iRODS logical file path in the format of /iRODS_zone/home/iRODS_user_name/file_path, which HydroShare would not have access to, so the referenced file cannot be downloaded. How should we encode the URI for such referenced file then?

**Comment by horsburgh** [comment]
@hyi - I may still be a little confused by this use case. I **think** what you are describing is creating a separate instance of HydroShare that will contain metadata-only records. There will be "resources" in that HydroShare instance, each will describe a dataset, but the files will not be stored in or accessible to the HydroShare instance and the users will not be able to access the data through that HydroShare instance. Is that right? 

HydroShare's resource data model, and in particular the resource map document and the science metadata document, were designed to use RDF/XML encoding. RDF uses URIs to identify resources (using RDF's terminology a **resource** can be a file) and their properties. In our HydroShare case, we are using URLs as URIs - which is great because that means there is an actual object at the other end of that URL that is either the file being described or a term from a vocabulary that has a definition in a HTML document. And, because the URL serves as both a name and a locator, those URLs can be resolved to return the individual files in the resource. Using URLs as the URIs is really useful in a "linked data" context because the "links" (the URLs) point at live resources (files that can be retrieved on the Internet).

In your example, if the files are never to be made available to the users for download, then any URI that is unique would work to identify each of the files in the resource, right? In this case, the URI acts only as a name and not as a locator. So, the HydroShare instance would be capable of listing all of the files in a resource, but would not be able to retrieve them. 

So - the short answer is - any URI that uniquely identifies/names a file could be used. This could be the iRODs logical file path - but keep in mind it will be encoded in the resource map and the science metadata document, so make sure it is something you want people to see.

But - that said - for this use case, will you be scripting/automatically building all of the resources from the files in the iRODS zone? This instance of HydroShare would essentially be read-only for users right?  If that's the case, I don't think this is the best example to be driving the design/implementation of a generalized reference resource for HydroShare (which I think should use URLs as URIs).

**Comment by hyi** [comment]
@horsburgh Thanks for the detailed thoughts. We have been discussing and refining the SC17 demo workflow since Friday, and it turns out we would most likely not need to create a reference file type in HydroShare after all. What you described above is in line with my initial thought, on HydroShare side, a resource holds metadata-only records and reference path to resource files that cannot be downloaded directly, however, the referenced files would be available in another cyberinfrastructure which would be demoed for SC17 and transferred to the on-demand provisioned appliances for users to do analysis and computation on the referenced data. However, this reference model may very likely be replaced by another model currently in discussion, so I will not continue with this reference file path at least for now. What your described above will serve as useful reference when we design a generalized reference resource for HydroShare regardless, so thanks for the detailed input on this.

**Comment by hyi** [comment]
@horsburgh @dtarb I have implemented a working prototype to add iRODS file reference functionalities for another project that build customized functionalities on top of HydroShare code base. Rather than creating a new resource type or file type, I ended up adding this iRODS file reference as a generally available functionality that is readily available across the board as part of iRODS file browser. Specifically, I added a checkbox as shown in the screen capture below:
![irods_file_selection](https://user-images.githubusercontent.com/476302/36811334-64b3636a-1c9b-11e8-80e0-7b866b98dc97.png)
When this checkbox is not checked by default, it functions as what we have now. But when it is checked, the selected iRODS files will be brought in as reference files added into whatever types of the resource they are added to. In the meantime, all iRODS reference file metadata are also brought in as extra_metadata key-value pairs in HydroShare. The following two screen captures show what the file content and extra metadata interface look like for referenced files in a resource:
![extra_metadata_avus](https://user-images.githubusercontent.com/476302/36812641-cb34f988-1c9f-11e8-8ab1-fc3cbaf637cd.png)
![ref_file_content](https://user-images.githubusercontent.com/476302/36812642-cb42403e-1c9f-11e8-8b60-e1f953158b9c.png)

Although this is specific for iRODS files, I can see it would apply to other file reference use cases discussed here. Want to keep you posted on progress with this and get your feedback as well on whether you think this is useful for HydroShare. Looping in @rayi113 @mjstealey @cbcunc  to keep them posted as well.
